ninio-snmp/src/main/java/com/davfx/ninio/snmp/EncryptionEngine.java

```diff
@@ -93,25 +100,70 @@ public byte[] regenerateKey(byte[] id, String password) {
 				messageDigest.update(passwordBytes, 0, l);
 				count += l;
 			}
-	
+
 			byte[] digest = messageDigest.digest();
-	
+
 			messageDigest.reset();
 			messageDigest.update(digest);
 			messageDigest.update(id);
 			messageDigest.update(digest);
-			d = messageDigest.digest();
-			cache.put(k, d);
+			key = messageDigest.digest();
+
+			if (privateKey && privacyProtocol != null) {
+				if (key.length >= privacyProtocol.minKeyLength()) {
+					if (key.length > privacyProtocol.maxKeyLength()) {
+						// truncate key
+						byte[] truncatedKey = new byte[privacyProtocol.maxKeyLength()];
+						System.arraycopy(key, 0, truncatedKey, 0, privacyProtocol.maxKeyLength());
+						key = truncatedKey;
+					}
+				} else {
+					// extend key if necessary
+					key = extendShortKey(key);
+				}
+			}
+
+			cache.put(k, key);
 			LOGGER.trace("Key regenerated");
 		}
-		return d;
+		return key;
+	}
+
+	public byte[] extendShortKey(byte[] shortKey) {
+		// we have to extend the key, currently only the AES draft
+		// defines this algorithm, so this may have to be changed for other
+		// privacy protocols
+		byte[] extKey = new byte[privacyProtocol.minKeyLength()];
+		int length = shortKey.length;
+		System.arraycopy(shortKey, 0, extKey, 0, length);
+
+		while (length < extKey.length) {
+			byte[] hash = hash(extKey, 0, length);
+
+			if (hash == null) {
+				return null;
+			}
+			int bytesToCopy = extKey.length - length;
+			if (bytesToCopy > authDigestAlgorithm.digestLength()) {
+				bytesToCopy = authDigestAlgorithm.digestLength();
+			}
+			System.arraycopy(hash, 0, extKey, length, bytesToCopy);
+			length += bytesToCopy;
+		}
+		return extKey;
 	}
-	
+
+	public byte[] hash(byte[] data, int offset, int length) {
```

Comment:
this `hash()` may be private since it's not used outside of the class ?
+ There are two similar `hash()` methods, if at least one is public and the other one is private we can quickly identify which one is used by other components.